planner/core/preprocess.go

```diff
@@ -1759,3 +1763,18 @@ func (p *preprocessor) initTxnContextProviderIfNecessary(node ast.Node) {
 		InfoSchema: p.ensureInfoSchema(),
 	})
 }
+
+func (p *preprocessor) tryAutoConvert(colDef *ast.ColumnDef) bool {
+	sessVars := p.ctx.GetSessionVars()
+	if !sessVars.SQLMode.HasStrictMode() && colDef.Tp.Tp == mysql.TypeVarchar &&
+		colDef.Tp.Charset != "" {
+		colDef.Tp.Tp = mysql.TypeBlob
```

Comment:
It seems that we can remove this line. I pass `TestLargeVarcharAutoConv` after remove it.

Reply:
I thought that sometimes charset can not resolved until `ddl.CreateTable`, so I add a check here. But I found this https://github.com/pingcap/tidb/blob/e389170624e227068213f523b7bed837ae3d95b1/planner/core/preprocess.go#L1254-L1260
So I think `colDef.Tp.Charset != "" ` can be removed now, but strict mode and column type may still need to be checked.

Reply:
Is there a misunderstanding here? I mean to remove `coldef.tp.tp = mysql.typeblob`

Reply:
Sorry for the misunderstanding. IMHO it may not be removed. 
```
mysql> create table i30328_1(a varbinary(70000), b varchar(70000000));
Query OK, 0 rows affected, 3 warnings (0.01 sec)

mysql> show warnings;
+---------+------+----------------------------------------------+
| Level   | Code | Message                                      |
+---------+------+----------------------------------------------+
| Warning | 1246 | Converting column 'a' from VARBINARY to BLOB |
| Warning | 1246 | Converting column 'a' from VARBINARY to BLOB |
| Warning | 1246 | Converting column 'b' from VARCHAR to TEXT   |
+---------+------+----------------------------------------------+
3 rows in set (0.00 sec)
```
If `colDef.Tp.Tp = mysql.TypeBlob` is removed, type convertion is not performed during preprocess stage, so convertion will happen and warning will be throw again in `checkTooBigFieldLengthAndTryAutoConvert`.

Reply:
Is there a test for this scenario? I can pass the test when I delete this line. If there is no test please help to make up.

Reply:
Yes, it is `(*testDBSuite5).TestTooBigFieldLengthAutoConvert` in `ddl/db_test.go`
If this line is removed, it will error as follows:
```
db_test.go:7833:
    tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1246 Converting column 'a' from VARBINARY to BLOB", "Warning 1246 Converting column 'b' from VARCHAR to TEXT"))
/home/chen/tidb/util/testkit/testkit.go:67:
    res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)
... obtained string = "" +
...     "[Warning 1246 Converting column 'a' from VARBINARY to BLOB]\n" +
...     "[Warning 1246 Converting column 'a' from VARBINARY to BLOB]\n" +
...     "[Warning 1246 Converting column 'b' from VARCHAR to TEXT]\n"
... expected string = "" +
...     "[Warning 1246 Converting column 'a' from VARBINARY to BLOB]\n" +
...     "[Warning 1246 Converting column 'b' from VARCHAR to TEXT]\n"
... sql:show warnings, args:[]

```